pandas/core/series.py

```diff
@@ -1043,7 +1046,15 @@ def _set_with_engine(self, key, value):
         # fails with AttributeError for IntervalIndex
         loc = self.index._engine.get_loc(key)
         validate_numeric_casting(self.dtype, value)
-        self._values[loc] = value
+        dtype, _ = infer_dtype_from_scalar(value)
+        if is_dtype_equal(self.dtype, dtype):
+            self._values[loc] = value
+        else:
+            # This only raises when index contains tuples
+            try:
+                self.loc[key] = value
```

Comment:
_set_with_engine is a fastpath, really shouldn't be going through self.loc

the whole infer_dtype_from_scalar stuff seems like it should be handled as part of validate_numeric_casting

the issue this PR is addressing is only for when we are indexing an entire column, right?